zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java

```diff
@@ -432,46 +475,103 @@ private void saveToFile() throws IOException {
-          if (registeredInterpreter.isDefaultInterpreter()) {
-            interpreterInfos.add(0,
-                new InterpreterSetting.InterpreterInfo(
-                    registeredInterpreter.getClassName(), registeredInterpreter.getName()));
-          } else {
-            interpreterInfos.add(new InterpreterSetting.InterpreterInfo(
-                registeredInterpreter.getClassName(), registeredInterpreter.getName()));
+    Preconditions.checkNotNull(refName, "name should not be null");
+    Preconditions.checkNotNull(interpreterInfos, "interpreterInfos should not be null");
+    Preconditions.checkNotNull(dependencies, "dependencies should not be null");
+    Preconditions.checkNotNull(option, "option should not be null");
+    Preconditions.checkNotNull(properties, "properties should not be null");
+//    Preconditions.checkNotNull(path, "path should not be null");
+
+    InterpreterSetting interpreterSetting;
+
+    synchronized (interpreterSettingsRef) {
+      if (interpreterSettingsRef.containsKey(refName)) {
+        interpreterSetting = interpreterSettingsRef.get(refName);
+
+        // Append InterpreterInfo
+        List<InterpreterInfo> infos = interpreterSetting.getInterpreterInfos();
+        boolean hasDefaultInterpreter = findDefaultInterpreter(infos);
+        for (InterpreterInfo interpreterInfo : interpreterInfos) {
+          if (!infos.contains(interpreterInfo)) {
+            if (!hasDefaultInterpreter && interpreterInfo.isDefaultInterpreter()) {
+              hasDefaultInterpreter = true;
+              infos.add(0, interpreterInfo);
+            } else {
+              infos.add(interpreterInfo);
+            }
           }
         }
-      }
 
-      InterpreterSetting intpSetting = new InterpreterSetting(name, groupName, interpreterInfos,
-          properties, dependencies, option);
+        // Append dependencies
+        List<Dependency> dependencyList = interpreterSetting.getDependencies();
+        for (Dependency dependency : dependencies) {
+          if (!dependencyList.contains(dependency)) {
+            dependencyList.add(dependency);
+          }
+        }
 
-      if (dependencies.size() > 0) {
-        loadInterpreterDependencies(intpSetting);
+        // Append properties
+        Properties interpreterProperties = interpreterSetting.getProperties();
+        for (String key : properties.stringPropertyNames()) {
+          if (!interpreterProperties.containsKey(key)) {
+            interpreterProperties.setProperty(key, properties.getProperty(key));
+          }
+        }
+
+      } else {
+        interpreterSetting = new InterpreterSetting(refName, null, interpreterInfos, properties,
+            dependencies, option, path);
+        interpreterSettingsRef.put(refName, interpreterSetting);
       }
+    }
 
-      intpSetting.setInterpreterGroupFactory(this);
-      interpreterSettings.put(intpSetting.id(), intpSetting);
-      saveToFile();
```

Comment:
I've replace the function of `add` to add `interpreterSettingRef` and add new method called `createNewSetting`. Thus this method is called when `InterpreterFactory.init()` only and init method run `saveToFile()` at the end of method. Thus It would be redundant. But, I've missed `saveToFile()` at the end of `createNewSetting`. I'll add it into `createNewSetting`